examples/credential-registry/src/lib.rs

```diff
@@ -0,0 +1,1292 @@
+use core::fmt::Debug;
+
+#[derive(Serialize, SchemaType)]
+struct RegisterCredentialEvent {
+    credential_id: Uuidv4,
+    holder_id:     PublicKeyEd25519,
+    schema_ref:    String,
+}
+
+#[derive(Serialize, SchemaType)]
+struct UpdateCredentialEvent {
+    credential_id: Uuidv4,
+    holder_id:     PublicKeyEd25519,
+    schema_ref:    String,
+}
+
+#[derive(Serialize, SchemaType)]
+enum Revoker {
+    Issuer,
+    Holder,
+    Other(PublicKeyEd25519),
+}
+
+#[derive(Serialize, SchemaType)]
+struct RevokeCredentialEvent {
+    credential_id: Uuidv4,
+    holder_id:     PublicKeyEd25519,
+    revoker:       Revoker,
+    reason:        Option<String>,
+}
+
+/// The type for a credential identifier.
+/// The uuidv4 identifier is generated externally by the issuer.
+type Uuidv4 = u128;
+
+#[derive(Serialize, SchemaType, PartialEq, Eq, Clone, Copy, Debug)]
+pub enum CredentialStatus {
+    Active,
+    Revoked,
+    Expired,
+    NotActivated,
+}
+
+/// Public verifiable credential data
+#[derive(Serialize, SchemaType, PartialEq, Eq, Clone, Debug)]
+pub struct CredentialData {
+    holder_id:            PublicKeyEd25519,
+    is_holder_revocable:  bool,
+    commitment:           Vec<u8>,
+    schema:               String,
+    serialization_schema: Vec<String>,
+    valid_from:           Option<Timestamp>,
+    valid_until:          Option<Timestamp>,
+}
+
+/// Public verifiable credential data, revocation flag and nonce
+#[derive(Serialize, SchemaType, PartialEq, Eq, Clone, Debug)]
+pub struct CredentialEntry {
+    credential_data:  CredentialData,
+    /// The nonce is used to avoid replay attacks when checking the holder's
+    /// signature on a revocation message.
+    revocation_nonce: u64,
+    is_revoked:       bool,
+}
+
+impl CredentialEntry {
+    /// Compute the credential status based on validity dates and the revocation
+    /// flag. If a VC is revoked the status will be `Revoked` regardless the
+    /// validity dates.
+    fn get_status(&self, now: Timestamp) -> CredentialStatus {
+        if self.is_revoked {
+            return CredentialStatus::Revoked;
+        }
+        if self.credential_data.valid_until.map_or(false, |x| x < now) {
+            return CredentialStatus::Expired;
+        }
+        if self.credential_data.valid_from.map_or(false, |x| now < x) {
+            return CredentialStatus::NotActivated;
+        }
+        CredentialStatus::Active
+    }
+}
+
+/// The registry state.
+// NOTE: keys are stored in a map, so one can refer to the keys by "names" in this case represented
+// by numbers. The keys could be removed and added, but external references (e.g. in DIDs) should
+// still be valid (unless a key was deliberatly removed)
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+pub struct State<S> {
+    issuer_metadata: MetadataUrl,
+    issuer_keys:     StateMap<u8, PublicKeyEd25519, S>,
+    revocation_keys: StateMap<u8, (PublicKeyEd25519, u64), S>,
```

Comment:
Looking more at the logic, is it really reasonable to hard-code this to u8? It seems like this set of revokers could in general be rather large.

Reply:
I woudn't expect too many revokers, but maybe it can be `u16` just in case

Reply:
If I understand correctly, we decided to keep it `u8`.